ext/autoload_php_files.c

```diff
@@ -0,0 +1,329 @@
+            }
+
+            dd_load_file(ZSTR_VAL(class_name));
+            if ((ce = zend_hash_find_ptr(EG(class_table), lc_name))) {
+                return ce;
+            }
+        }
+
+        if (get_DD_TRACE_OTEL_ENABLED() && zend_string_starts_with_literal(lc_name, "opentelemetry\\") && !DDTRACE_G(otel_is_loaded)) {
+            DDTRACE_G(otel_is_loaded) = 1;
+            dd_load_files("opentelemetry");
+            if ((ce = zend_hash_find_ptr(EG(class_table), lc_name))) {
+                return ce;
+            }
+        }
+    }
+#if PHP_VERSION_ID >= 80000
+    if (EXPECTED(dd_prev_autoloader != NULL)) {
+        return dd_prev_autoloader(class_name, lc_name);
+    }
+#endif
+    return NULL;
+}
+
+#if PHP_VERSION_ID < 80000
+ZEND_TLS bool dd_in_autoload = false;
+ZEND_TLS bool dd_has_registered_spl_autoloader = false;
+
+static inline bool dd_legacy_autoload_wrapper(INTERNAL_FUNCTION_PARAMETERS) {
+    UNUSED(return_value);
+    zend_string *class_name;
+
+    if (dd_in_autoload) {
+        return false;
+    }
+
+    ZEND_PARSE_PARAMETERS_START_EX(ZEND_PARSE_PARAMS_QUIET, 1, 1)
+        Z_PARAM_STR(class_name)
+    ZEND_PARSE_PARAMETERS_END_EX(return false;);
+
+    zend_string *lower = zend_string_tolower(class_name);
+    bool found = dd_perform_autoload(class_name, lower) != NULL;
+    zend_string_release(lower);
+
+    return found;
+}
+
+static ZEND_NAMED_FUNCTION(dd_wrap_autoload_register_fn) {
+    dd_spl_autoload_register_fn_handler(INTERNAL_FUNCTION_PARAM_PASSTHRU);
+    if (!EG(exception)) {
+        dd_has_registered_spl_autoloader = true;
+    }
+}
+
+static ZEND_NAMED_FUNCTION(dd_wrap_autoload_unregister_fn) {
+    dd_spl_autoload_unregister_fn_handler(INTERNAL_FUNCTION_PARAM_PASSTHRU);
+    ddtrace_autoload_rinit();
+}
+
+static ZEND_NAMED_FUNCTION(dd_perform_autoload_fn) {
+    if (!dd_legacy_autoload_wrapper(INTERNAL_FUNCTION_PARAM_PASSTHRU)) {
+        dd_spl_autoload_fn_handler(INTERNAL_FUNCTION_PARAM_PASSTHRU);
+    }
+}
+
+static ZEND_NAMED_FUNCTION(dd_perform_autoload_call_fn) {
+    if (!dd_legacy_autoload_wrapper(INTERNAL_FUNCTION_PARAM_PASSTHRU)) {
+        dd_in_autoload = true; // spl_autoload_call may call spl_autoload
+        dd_spl_autoload_call_fn_handler(INTERNAL_FUNCTION_PARAM_PASSTHRU);
+        dd_in_autoload = false;
+    }
+}
+
+static ZEND_NAMED_FUNCTION(dd_wrap_autoload_functions_fn) {
+    if (!dd_has_registered_spl_autoloader) {
+        EG(autoload_func) = dd_spl_autoload_fn;
+    }
+
+    dd_spl_autoload_functions_fn_handler(INTERNAL_FUNCTION_PARAM_PASSTHRU);
+
+    EG(autoload_func) = dd_spl_autoload_call_fn;
+}
+
+void ddtrace_autoload_rinit(void) {
+    if (!EG(autoload_func)) {
+        dd_has_registered_spl_autoloader = false;
+        EG(autoload_func) = dd_spl_autoload_call_fn;
+    }
+    dd_in_autoload = false;
+}
+#endif
+
+void ddtrace_autoload_minit(void) {
+#if PHP_VERSION_ID >= 80000
+    dd_prev_autoloader = zend_autoload;
+    zend_autoload = dd_perform_autoload;
```

Comment:
Well, we call the previous autoloader. And yes, ti would fail if other extensions replace it without calling the previous autoloader. But who would do that :-D